components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1359,6 +1359,73 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
                           mojom::ProviderError::kSuccess, "");
 }
 
+void JsonRpcService::GetERC721TokenURI(const std::string& contract_address,
+                                       const std::string& token_id,
+                                       const std::string& chain_id,
+                                       GetERC721TokenURICallback callback) {
+  auto network_url = GetNetworkURL(prefs_, chain_id, mojom::CoinType::ETH);
+  if (!network_url.is_valid()) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  if (!EthAddress::IsValidAddress(contract_address)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  uint256_t token_id_uint = 0;
+  if (!HexValueToUint256(token_id, &token_id_uint)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  std::string data;
+  if (!erc721::TokenURI(token_id_uint, &data)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetERC721TokenURI,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  RequestInternal(
+      eth::eth_call("", contract_address, "", "", "", data, "latest"), true,
+      network_url, std::move(internal_callback));
+}
+
+void JsonRpcService::OnGetERC721TokenURI(
+    GetERC721TokenURICallback callback,
+    const int status,
+    const std::string& body,
+    const base::flat_map<std::string, std::string>& headers) {
+  if (status < 200 || status > 299) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  std::string result;
+  if (!eth::ParseEthCall(body, &result)) {
+    mojom::ProviderError error;
+    std::string error_message;
+    ParseErrorResult<mojom::ProviderError>(body, &error, &error_message);
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
```

Comment:
Right now, the code doesn't attempt to decode the response. For example, calls to `tokenURI` on contract `0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d` will return
```
0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000037697066733a2f2f516d65536a53696e4870506e6d586d73704d6a776958794e367a533445397a63636172694752336a7863615774712f31000000000000000000
```
rather than
```
ipfs://QmeSjSinHpPnmXmspMjwiXyN6zS4E9zccariGR3jxcaWtq/1
```

I saw examples of both in the code base.  For example, we don't decode the eth_call responses for GetERC20TokenBalance:
https://github.com/brave/brave-core/blob/9e1daf1aad4360e5248bf856719429b1f0093005/components/brave_wallet/browser/json_rpc_service_unittest.cc#L1300-L1314

But it looks like we do make an attempt for calls to GetERC721OwnerOf which return an address:
https://github.com/brave/brave-core/blob/9e1daf1aad4360e5248bf856719429b1f0093005/components/brave_wallet/browser/json_rpc_service_unittest.cc#L1804-L1818

I followed the model in GetERC20TokenBalance and didn't decode since that's also using eth_call.  But I could see an argument for fully decoding before sending the result to frontends.

Reply:
I think backend should be responsible of parsing the result for sharing it across all platforms.